### DISCUSSION TITLE
Initiative: reduce amount of dependencies

### DISCUSSION THREAD

**Discussion by AbhiPrasad** [semantic-release/semantic-release#3376] [Ideas]
### New feature motivation

[`semantic-release`](https://npmgraph.js.org/?q=semantic-release) has quite the dependency graph, loading 110 MB / 493 dependencies. It would be nice if we could reduce the number of dependencies here. 

I'm starting this as a discussion to get feedback from the maintainers, but I'm happy to make PRs and contribute these changes myself.

### New feature description

The package engines does now require `>=20.8.1`. I think that means we can remove a lot of packages that support older node versions, but bring in polyfills/extra code to do so.

### New feature implementation

Some areas we can improve:

- `git-log-parser`: https://npmgraph.js.org/?q=git-log-parser@1.2.0, we can probably remove `traverse` requirement with newer node features
- `https://github.com/semantic-release/npm`: https://npmgraph.js.org/?q=@semantic-release/npm@12.0.1, remove usage of `tempy`, simplify `package.json` reading

**Comment by gr2m** [comment]
Hey @AbhiPrasad, I think the idea is great and if we could reduce the amount size of our dependency try without compromising security, reliability and performance I think we are very open to it. I'd suggest to send many smaller pull requests, it will make it much easier for us to discuss and review

**Comment by AbhiPrasad** [comment]
One area of improvement we identified was `read-pkg`, which we can switch to `read-package-json-fast`: https://github.com/43081j/ecosystem-cleanup/issues/20

> (read-package-json-fast) is near enough 10x smaller, is maintained by the NPM team directly, and is more standard because of that.

PR: https://github.com/semantic-release/npm/pull/745

**Reply by benmccann** [reply]
Here's another location that could be updated: https://github.com/semantic-release/release-notes-generator/blob/7d8ea2ce749de931c8a87a45ffdf7b800a1a3301/index.js#L8

**Comment by AbhiPrasad** [comment]
Another area of improvement was modernizing `git-log-parser`, which `semantic-release` relies on.

@valadaptive kindly went ahead and made up-stream-fixes to make this pull in less deps, which was released with `1.2.1`: https://www.npmjs.com/package/git-log-parser/v/1.2.1

This now makes semantic-release have 94 MB / 427 deps 🔥🔥

**Reply by gr2m** [reply]
awesome, thank you @valadaptive 💐

**Comment by AbhiPrasad** [comment]
Given we are using a more modern Node.js version, we can switch to using the built-in `AggregateError` instead of pulling in a package.

I opened three PRs to do this:
- https://github.com/semantic-release/github/pull/866
- https://github.com/semantic-release/npm/pull/831
- https://github.com/semantic-release/semantic-release/pull/3377

In the [dependency graph](https://npmgraph.js.org/?q=semantic-release) we still have `aggregate-error` relied on by [npm@10.8.1](https://npmgraph.js.org/?q=npm@10.8.1), but I'll also work toward cleaning that upstream.

**Comment by benmccann** [comment]
Probably the largest thing would be to remove `npm` as a dependency. It pulls in 185 packages. It shouldn't be necessary to have it as a dependency because it should be installed on the machine already so that you can call out to it as a separate process. This is what changesets does: https://github.com/changesets/changesets/blob/b59375614b1b3dabdf67806cd202defb314686a8/packages/cli/src/commands/publish/npm-utils.ts#L163

**Reply by gr2m** [reply]
We had this discussion in the past, let's leave that out of the scope of this initiative, it's complicated.

**Comment by benmccann** [comment]
`issue-parser` uses five different `lodash` libraries

**Comment by benmccann** [comment]
https://github.com/klaudiosinani/signale/blob/64cea86a6f4a5b4509fe25550dd1ce8f058c9dcc/src/signale.js#L7 uses a deprecated package. The new one is lighter

It also uses an outdated version of chalk

**Comment by gr2m** [comment]
hey sorry I didn't get back to your PRs yet. I've been out on vacation and sick now. I didn't forget and appreciate your efforts
